README.md
# Coeditor

Yet Another Pair Programming Package for Atom.

## Demo
_Click to view high quality video on YouTube_

[![demo](demo.gif)](https://youtu.be/EOCw2HfRxZE)

## Usage
This package requires a server to forward changes to each clients.
Currently there is no available common server yet, but it is quite simple to deploy one by your own.

_prerequisite: node.js (v>=6.10.1)_

+ Clone this repo
  - `git clone https://github.com/nicktogo/coeditor.git`  
+ Install dependencies
  - `cd coeditor`
  - `npm install`
+ Start Server
  - `npm start`
  - By default, the server will listen on port 9090
+ Toggle package in Atom
  - ~~open `coeditor` folder in Atom~~
  - ~~change `address` in `coeditor.js` and save~~
  - ~~client can now set server address when toggling package~~
  - ~~link package to Atom~~
    - ~~`cd coeditor`~~
    - ~~`apm link`~~
  - ~~reload Atom, `ctrl + shift + F5`~~
  - install coeditor via Atom
  - modify settings in the package's settings view firstly
    - server address (with port)
    - client id which will show in indicator
    - indicator background color
    - line ending
  - open project folder in Atom
  - toggle coeditor, `ctrl + shift + p` then type `toggle`, hit `coeditor: toggle`
  - input ~~server address,~~ session id ~~and client id~~
  - check your server console
  - disconnect, `ctrl + shift + p` then type `disconnect`, hit `coeditor: disconnect`

## Features
+ Non-conflict collaborative text editing
+ Text highlight synchronization
+ Collaborative awareness
+ Project sync
+ File grammar sync
+ Active tab sync
+ Project's files saving and closing sync

## Known issue
+ If clients use different types for line ending (i.e. CRLF or LF) with each other, this package will cause files end up with inconsistent content.
**MAKE SURE ALL CLIENTS USE THE SAME TYPE FOR LINE ENDING AND SET TYPE IN THIS PACKAGE'S SETTING VIEW! (EITHER CRLF OR LF WILL DO)**

## TODO
+ ~~Customized configuration (Server address, random session id or given by user)~~ DONE
+ ~~Broadcast disconnections to other living Atom clients~~ DONE
+ ~~Check socket status before using~~ DONE
+ ~~Broadcast server close to each Atom clients, and disconnect all clients~~ DONE
+ ~~Text selection background and cursor colors for different Atom clients ?~~ DONE  
+ Suggestions Synchronization NO API
+ ~~Settings view~~ DONE
+ ~~Working folder files synchronization~~ DONE
+ ~~File operations synchronization~~ DONE
  - ~~save -- DONE, but limited, don't work for new file which does not have path~~
  - ~~close~~
+ ~~Tab sync~~
+ add sync tab switcher (via command or configuration)
+ Heartbeat between client and server
+ Build cloud server
+ Write spec

## Tentative
+ ~~Working folder files synchronization~~ MOVED TO TODO list
+ Server uses in-memory storage when running, so no file content will be stored in server when it is shutdown, might change to MongoDB

## Thanks
+ Dr. Fan Hongfei
+ [ShareDB](https://github.com/share/sharedb)
+ [ws](https://github.com/websockets/ws)


CHANGELOG.md
## 0.1.0 - First Release
* Every feature added
* Every bug fixed

## 0.2.4
* fixed [#3](https://github.com/nicktogo/coeditor/issues/3)


spec/coeditor-spec.js
'use babel';

import Coeditor from '../lib/coeditor';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('Coeditor', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('coeditor');
  });

  describe('when the coeditor:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.coeditor')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'coeditor:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.coeditor')).toExist();

        let coeditorElement = workspaceElement.querySelector('.coeditor');
        expect(coeditorElement).toExist();

        let coeditorPanel = atom.workspace.panelForItem(coeditorElement);
        expect(coeditorPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'coeditor:toggle');
        expect(coeditorPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.coeditor')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'coeditor:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let coeditorElement = workspaceElement.querySelector('.coeditor');
        expect(coeditorElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'coeditor:toggle');
        expect(coeditorElement).not.toBeVisible();
      });
    });
  });
});


spec/coeditor-view-spec.js
'use babel';

import CoeditorView from '../lib/coeditor-view';

describe('CoeditorView', () => {
  it('has one valid test', () => {
    expect('life').toBe('easy');
  });
});


lib/router.js
'use babel';

import ShareDB from 'sharedb/lib/client';
import WebSocket from 'ws';
import EventHandler from './EventHandler';
import path from 'path';

/*
* The Router is the central processor, it is responsible for several things:
* 1. managing WebSocket, which is used to communicate with server.
* 2. managing EventHandlers, which correspond to Atom's TextEditor.
* 3. responsing to system-wide events.
* 4. dispatch editor's events to EventHandlers.
*/

export default class Router {
  constructor(sessionId, coeditor) {

    this.coeditor = coeditor;
    this.sessionId = sessionId;
    this.serverAddress = atom.config.get('coeditor.serverAddress');
    this.clientId = atom.config.get('coeditor.clientId');

    // filepath -> EventHandler
    this.eventHandlerMap = new Map();

    // TODO move sync tab option to config
    this.syncTab = true;

    this.syncTabOn = true;

    this.socket = new WebSocket('ws://' + this.serverAddress);

    this.socket.on('open', () => {
      let initData = {
        a         : 'meta',
        type      : 'init',
        sessionId : this.sessionId,
        clientId  : this.clientId
      };
      this.sendSocketMsg(JSON.stringify(initData));

      this.connection = new ShareDB.Connection(this.socket);

      this.coeditor.onSocketOpened();
      atom.notifications.addSuccess('Connected to: ' + this.socket.url);
    });

    this.socket.on('message', (msg) => {
      let data  = JSON.parse(msg);

      // an op msg for ShareDB, do not process it
      if (data.a !== 'meta') {
        return;
      }

      // system-wide event
      // TODO move to coeditor.js
      if (data.type === 'activePaneItemChanged' && this.syncTabOn) {
        this.syncTab = false;
        let targetUri = path.join(this.coeditor.projectPath, data.uri);
        atom.workspace.open(targetUri).then( () => {
          this.syncTab = true;
        });
        return;
      }

      // system-wide event
      // TODO move to coeditor.js
      if (data.type === 'editorClosed') {
        let targetPath = data.path;
        atom.workspace.getPaneItems().forEach( (item) => {
          let itemPath = item.getPath();
          let paths = atom.project.relativizePath(itemPath);
          paths[1] = this.coeditor.normalizePath(paths[1]);
          if (typeof itemPath !== 'undefined' && paths[1] === targetPath) {
            let pane = atom.workspace.paneForItem(item);
            pane.destroyItem(item);
            return;
          }
        });
        return;
      }

      // system-wide event
      if (data.type === 'socketClose') {
        atom.notifications.addInfo(data.clientId + ' left.');
        for (let handler of this.eventHandlerMap.values()) {
          handler.resetMarker(data.clientId);
        }
        return;
      }

      // editor event
      let targetPath = data.path;
      let targetHandler = this.eventHandlerMap.get(targetPath);
      if (typeof targetHandler !== 'undefined') {
        targetHandler.on(msg);
      }
    });

    this.socket.on('close', (code, reason) => {
      this.coeditor.disconnect();
    });
  }

  addHandler(path, editor) {
    let doc = this.connection.get(this.sessionId, path);
    let handler = new EventHandler(doc, editor, path, this);
    this.eventHandlerMap.set(path, handler);
    return handler.listen();
  }

  // send socket msg via Router's WebSocket
  sendSocketMsg(msg) {
    if (this.socket.readyState === WebSocket.OPEN) {
      this.socket.send(msg);
    }
  }

  destroy() {
    for (let handler of this.eventHandlerMap.values()) {
      handler.destroy();
    }
    if (this.socket) {
      this.socket.close();
      this.socket = null;
      atom.notifications.addSuccess("Disconnected");
    } else {
      atom.notifications.addWarning("No active socket/connection");
    }
  }
}


lib/coeditor.js
'use babel';

import CoeditorView from './coeditor-view';
import { CompositeDisposable, Range } from 'atom';
import ShareDB from 'sharedb/lib/client';
import otText from 'ot-text';
import WebSocket from 'ws';
import crypto from 'crypto';
import path from 'path';
import EventHandler from './EventHandler';
import Router from './router';

ShareDB.types.register(otText.type);

export default {

  // config of this package
  config: {
    serverAddress: {
      type: 'string',
      default: '192.168.1.175:9090',
      order: 1
    },
    clientId: {
      type: 'string',
      default: crypto.randomBytes(5).toString('hex'),
      order: 2
    },
    indicatorBackgroundColor: {
      type: 'color',
      default: 'rgba(31, 161, 93, 1)',
      order: 3
    },
    lineEnding: {
      type: 'string',
      default: 'CRLF',
      enum: ['CRLF', 'LF'],
      order: 4
    }
  },

  coeditorView: null,
  subscriptions: null,

  socket: null,
  push: null,
  doc: null,
  editor: null,
  buffer: null,
  marker: null,

  initialized: null,

  activate(state) {
    this.initial(state);
  },

  deactivate() {
    this.disconnect();
  },

  serialize() {
    return {
    };
  },

  initial(state) {
    if (this.initialized) {
      return;
    }

    this.coeditorView = new CoeditorView(state.viewState);
    // Events subscribed to in atom's system can be easily cleaned up with a CompositeDisposable
    this.subscriptions = new CompositeDisposable();

    // Register command that toggles this view
    atom.commands.add('atom-workspace', {
      'coeditor:toggle': () => this.toggle()
    });
    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'coeditor:disconnect': () => this.disconnect()
    }));
    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'coeditor:syncTab': () => this.syncTab()
    }));
    this.subscriptions.add(atom.commands.add(this.coeditorView.element, {
      'core:confirm': () => this.connect(),
      'core:cancel': () => this.coeditorView.close()
    }));
    this.initialized = true;
  },

  syncTab() {
    this.router.syncTabOn = !this.router.syncTabOn;
  },

  disconnect() {
    if (!this.initialized) {
      return;
    }
    this.coeditorView.destroy();
    this.subscriptions.dispose();
    this.router.destroy();
    this.initialized = false;
  },

  toggle() {
    this.initial({});
    this.coeditorView.show(crypto.randomBytes(20).toString('hex'));
  },

  connect() {
    this.coeditorView.close();
    let sessionId = this.coeditorView.sessionIdEditor.getText();
    this.router = new Router(sessionId, this);
  },

  // FIXME when there is no active TextEditor, this method will fail
  initProjectPath() {
    let activeEditor = atom.workspace.getActiveTextEditor();
    this.projectPath = atom.project.relativizePath(activeEditor.getPath())[0];
  },

  onSocketOpened() {
    this.initProjectPath();

    this.subscriptions.add(atom.workspace.observeTextEditors( (editor) => {
      let editorPath = editor.getPath();
      let paths = atom.project.relativizePath(editorPath);
      paths[1] = this.normalizePath(paths[1]);
      if (this.projectPath === paths[0]) {
        let disposables = this.router.addHandler(paths[1], editor);
        this.subscriptions.add(...disposables);
      }
    }));

    this.subscriptions.add(atom.workspace.onDidChangeActivePaneItem( (item) => {
      if (!this.router.syncTabOn) {
        return;
      }
      if (!this.router.syncTab) {
        return;
      }
      if (!atom.workspace.isTextEditor(item)) {
        return;
      }
      let result = this.isProjectFile(item.getPath());
      if (result.isProjectFile) {
        let data = {
          a: 'meta',
          type: 'activePaneItemChanged',
          uri: result.relativePath
        };
        setTimeout( () => {
          this.router.sendSocketMsg(JSON.stringify(data));
        }, 0);
      }
    }));
  },

  isProjectFile(absolutePath) {
    let result = {
      isProjectFile: false
    };
    let paths = atom.project.relativizePath(absolutePath);
    if (this.projectPath === paths[0]) {
      result.isProjectFile = true;
      result.relativePath = this.normalizePath(paths[1]);
      return result;
    }
    return result;
  },


  // convert \ in path to /
  normalizePath(pathStr) {
    if (pathStr == null) {
      return;
    }
    if (process.platform === 'win32') {
      return pathStr.split(path.sep).join("/");
    }
    return pathStr;
  }
};


server/server.js
var ShareDB = require('sharedb');
var WebSocket = require('ws');
var WebSocketStream = require('websocket-push-stream');
var otText = require('ot-text');

ShareDB.types.register(otText.type);

allSessions = [];

var backend = new ShareDB();
backend.use('op', (request, callback) => {
  callback();
  setTimeout( () => {
    let ws = request.agent.stream.ws;
    let cursors = allSessions[ws.sessionId].cursors;
    if (typeof cursors !== 'undefined') {
      console.log('Broadcasting ' + ws.clientId + '\'s cursors');
      for (let path in cursors) {
        if (cursors.hasOwnProperty(path) && JSON.parse(cursors[path]).clientId === ws.clientId) {
          console.log(path);
          broadcastMsg(cursors[path], ws);
        }
      }
      cursors = {};
    }
  }, 0);
});
startServer();

function startServer() {
  // Create a WebSocket Server
  // and connect any incoming WebSocket connection to ShareDB
  var wss = new WebSocket.Server({port: 9090}, () => {
    console.log('WebSocket Server Created.');
  });
  wss.on('connection', function(ws) {
    var stream = new WebSocketStream(ws);

    ws.on('message', function(msg) {
      let data = JSON.parse(msg);
      if (data.a === 'meta') {
        console.log('Received meta data:');
        console.log(data);
        console.log('\n');
        if (data.type === 'init') {
          // create or join a session
          ws.createOrJoinSession(data);
          ws.send(JSON.stringify(allSessions[ws.sessionId].tabs));
        } else {
          // tab changes: add or remove tab
          let logTabs = false;
          if (data.type === 'editorClosed') {
            let tabs = allSessions[ws.sessionId].tabs;
            let index = tabs.indexOf(data.path);
            if (index > -1) {
              tabs.splice(index, 1);
              console.log(data.path + ' removed.');
              logTabs = true;
            }
          } else if (data.type === 'addTab') {
            let tabs = allSessions[ws.sessionId].tabs;
            if (tabs.indexOf(data.uri) != -1) {
              return;
            }
            tabs.push(data.uri);
            logTabs = true;
            console.log(data.uri + ' added');
          } else if (data.type === 'cursorMoved') {
            let cursors = allSessions[ws.sessionId].cursors;
            cursors[data.path] = msg;
            return;
          }

          if (logTabs) {
            console.log('current tabs: ');
            console.log(allSessions[ws.sessionId].tabs);
            console.log('\n');
          }
          // other meta changes: cursor position, text selection
          // and open/save/close file
          broadcastMsg(msg, ws);
        }
      } else {
        // OT
        console.log(data);
        stream.push(JSON.parse(msg));
      }
    });

    ws.on('close', (code, reason) => {
      // socket client closed due to server closed, do not broadcast
      if (code === 1006) {
        return;
      }
      let index = allSessions[ws.sessionId].wss.indexOf(ws);
      if (index > -1) {
        allSessions[ws.sessionId].wss.splice(index, 1);
        console.log('We just lost one connection: ' + ws.clientId + ' from ' + ws.sessionId);
        console.log('Now ' + ws.sessionId + ' has ' + allSessions[ws.sessionId].wss.length + ' connection(s)');
        console.log('\n');
        let msg = {
          a: 'meta',
          type: 'socketClose',
          clientId: ws.clientId
        };
        broadcastMsg(JSON.stringify(msg), ws);
      }
    });

    backend.listen(stream);
    console.log('Got one connection...\n');
  });

  process.on('SIGINT', () => {
    wss.close( () => {
      process.exit();
    });
  });

};

function broadcastMsg(msg, ws) {
  let sockets = allSessions[ws.sessionId].wss;
  sockets.forEach( (socket) => {
    if (socket.readyState === WebSocket.OPEN && (socket.getId() !== ws.getId())) {
      console.log('Broadcasting msg to ' + socket.clientId + '\n');
      console.log(msg);
      console.log('\n');
      setTimeout( () => {
        socket.send(msg);
      }, 0);
    }
  });
}

WebSocket.prototype.createOrJoinSession = function(data) {
  let sessionId = data.sessionId;
  let clientId  = data.clientId;
  this.sessionId = sessionId;
  this.clientId  = clientId;
  if (typeof allSessions[sessionId] === 'undefined') {
    let session = {};
    session.wss = [];
    session.tabs = [];
    session.cursors = {};
    allSessions[sessionId] = session;
  }
  allSessions[sessionId].wss.push(this);
  console.log('Session ' + sessionId + ' adds ' + clientId + '\n');
};

WebSocket.prototype.getId = function() {
  return this.upgradeReq.headers['sec-websocket-key'];
};


lib/coeditor-view.js
'use babel';

import {TextEditor} from 'atom';

export default class CoeditorView {

  constructor(serializedState) {
    // Create root element
    this.element = document.createElement('div');
    this.element.classList.add('coeditor');

    // Create sessionId input element
    let sessionIdLabel = document.createElement('div');
    sessionIdLabel.textContent = 'Session Id';
    this.element.appendChild(sessionIdLabel);
    this.sessionIdEditor = new TextEditor({mini: true});
    this.element.appendChild(this.sessionIdEditor.element);
  }

  close() {
    if (this.panel.isVisible()) {
      this.panel.hide();
    }
  }

  show(sessionId) {
    if (!this.panel) {
      this.panel = atom.workspace.addModalPanel({item: this, visible: false});
    }
    this.sessionIdEditor.setText(sessionId);
    this.panel.show();
    this.sessionIdEditor.element.getModel().selectAll();
    this.sessionIdEditor.element.focus();
  }

  // Returns an object that can be retrieved when package is activated
  serialize() {}

  // Tear down any state and detach
  destroy() {
    this.element.remove();
  }

  getElement() {
    return this.element;
  }

}


server/WebSocketStream.js
var util = require('util');
var Duplex = require('stream').Duplex;

function WebSocketStream(ws) {
  // Make work with or without 'new'
  if (!(this instanceof WebSocketStream)) {
    return new WebSocketStream(ws);
  }

  Duplex.call(this, {objectMode: true});
  this.ws = ws;
  var self = this;
  ws.on('close', function() {
    self.push(null); // end readable stream
    self.end(); // end writable stream

    self.emit('close');
    self.emit('end');
  });

  this.on('error', function() {
    ws.close();
  });
  this.on('end', function() {
    ws.close();
  });
};
util.inherits(WebSocketStream, Duplex);

WebSocketStream.prototype._read = function() {};
WebSocketStream.prototype._write = function(chunk, encoding, next) {
  this.ws.send(JSON.stringify(chunk));
  next();
};

module.exports = WebSocketStream;


lib/EventHandler.js
'use babel';

import {Point, Range} from 'atom';

export default class EventHandler {
  constructor(doc, editor, path, router) {
    this.doc = doc;
    this.editor = editor;
    this.path = path;
    this.router = router;
    this.clientId = atom.config.get('coeditor.clientId');
    this.push = true;
    this.buffer = this.editor.getBuffer();
    this.indicatorMap = new Map();
    this.highlightMap = new Map();

    this.doc.subscribe( (error) => {
      if (error) throw error;
      if (this.doc.type === null) {
        this.doc.create(this.buffer.getText(), 'text');
      } else {
        let data = this.doc.data;
        this.modify( () => {
          this.buffer.setText(data);
        })
      }

      let data = {
        a: 'meta',
        type: 'addTab',
        uri: path // TextEditor's uri is it's absolute path, when sending data, we send relative path, then when we receive it, we will convert to absolute path
      };
      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);
    });
  }

  listen() {
    // remote OT events
    this.doc.on('op', (op, source) => {
      if (source) {
        return;
      }
      console.log('on OP:');
      console.log(op);

      var pos = 0;
      for (var i = 0; i < op.length; i++) {
        var component = op[i];
        switch (typeof component) {
          case 'number':
            pos += component;
            break;
          case 'string':
            this.onInsert(pos, component);
            pos += component.length;
            break;
          case 'object':
            this.onRemove(pos, component.d);
            break;
          default:
            break;
        }
      }
    });

    // local events
    let disposables = [];

    disposables.push(this.editor.onDidChangeGrammar( () => {

      let grammar = this.editor.getGrammar();
      let data = {
        a: 'meta',
        type: 'grammar',
        path: this.path,
        grammar: grammar.scopeName
      }

      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);
    }));

    disposables.push(this.editor.onDidSave( (event) => {
      if (!this.push) {
        return;
      }

      let data = {
        a: 'meta',
        type: 'save',
        path: this.path
        // path: atom.project.relativizePath(event.path)[1]
      };

      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);
    }));

    disposables.push(this.editor.onDidDestroy( () => {
      let data = {
        a: 'meta',
        type: 'editorClosed',
        path: this.path,
        clientId: this.clientId
      };

      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);

    }));

    disposables.push(this.editor.onDidChangeSelectionRange( (event) => {
      if (!this.push) {
        return;
      }
      let data = {
        a: 'meta',
        type: 'highlight',
        path: this.path,
        newRange: event.newBufferRange,
        clientId: this.clientId
      };

      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);
    }));

    disposables.push(this.editor.onDidChangeCursorPosition( (event) => {
      console.log('on did change cursor position');
      let data = {
        a: 'meta',
        type: 'cursorMoved',
        path: this.path,
        indicatorColor: atom.config.get('coeditor.indicatorBackgroundColor'),
        newPosition: event.newBufferPosition,
        clientId: this.clientId
      };

      setTimeout( () => {
        this.router.sendSocketMsg(JSON.stringify(data));
      }, 0);

      let position = this.buffer.characterIndexForPosition(event.newBufferPosition);
      let op = [position, {d: 0}];
      setTimeout( () => {
        this.doc.submitOp(op);
      }, 0);
    }));

    disposables.push(this.buffer.onDidChange((event) => {
      if (!this.push) {
        return;
      }
      console.log('on did change text');
      // remove text event
      if (event['oldText'].length !== 0) {
        let lineEnding = atom.config.get('coeditor.lineEnding');
        let oldText = lineEnding === 'CRLF' ?
                      event['oldText'].replace(/\r?\n/g, "\r\n") :
                      event['oldText'];
        console.log(oldText.length);
        let startPoint = event['oldRange'].start;
        let position = this.buffer.characterIndexForPosition(startPoint);
        let op = [position, {d: oldText.length}];
        this.doc.submitOp(op);
      }
      // insert text event
      if (event['newText'].length !== 0) {
        let startPoint = event['newRange'].start;
        let position = this.buffer.characterIndexForPosition(startPoint);
        let op = [position, event['newText']];
        this.doc.submitOp(op);
      }
    }));

    return disposables;
  }

  // TODO pass msg directly to hander function
  on(msg) {
    let data = JSON.parse(msg);
    switch (data.type) {
      case 'grammar':
        this.onGrammar(data.grammar);
        break;
      case 'cursorMoved':
        // setTimeout( () => {
          this.onCursorMoved(data.newPosition, data.clientId, data.indicatorColor);
        // }, 1000);
        break;
      case 'highlight':
        this.onHighlight(data.newRange, data.clientId);
        break;
      case 'save':
        this.onSave(data.path);
        break;
      default:
    }
  }

  onGrammar(grammar) {
    this.editor.setGrammar(atom.grammars.grammarForScopeName(grammar));
  }

  onCursorMoved(positionObj, clientId, indicatorColor) {
    // type = overlay
    let overlayMarker = this.indicatorMap.resetMarker(clientId);

    overlayMarker = this.editor.markBufferPosition(Point.fromObject(positionObj), {
      'invalidate': 'never'
    });

    this.indicatorMap.set(clientId, overlayMarker);

    if (overlayMarker.isDestroyed()) {
      return;
    }
    this.editor.decorateMarker(overlayMarker, {
      type: 'overlay',
      position: 'head',
      item: this.getIndicator(clientId, indicatorColor)
    });
    console.log('onCursorMoved: ');
    console.log(positionObj);
  }

  getIndicator(text, indicatorColor) {
    let indicatorDiv = document.createElement('div');
    Object.assign(indicatorDiv.style, {
      transform: 'translate(0, -100%)',
      display: 'flex'
    });

    let cursorDiv = document.createElement('div');
    Object.assign(cursorDiv.style, {
      alignItems: 'stretch',
      borderLeft: '2px solid ' + indicatorColor
      // borderLeft: '2px solid rgba(31, 161, 93, 1)'
    });
    cursorDiv.animate([
      {opacity: '1'},
      {opacity: '0'}
    ], {
      duration: 300,
      iterations: 'Infinity',
      direction: 'alternate'
    });
    indicatorDiv.appendChild(cursorDiv);

    let textDiv = document.createElement('div');
    textDiv.textContent = text;
    let style = {
      alignItems: 'stretch',
      backgroundColor: indicatorColor,
      paddingLeft: '1px',
      fontFamily: 'comic sans, comic sans ms, cursive, verdana, arial, sans-serif'
    }
    Object.assign(textDiv.style, style);
    indicatorDiv.appendChild(textDiv);

    return indicatorDiv;
  }

  onHighlight(rangeObj, clientId) {

    let marker = this.highlightMap.resetMarker(clientId);

    if (Point.fromObject(rangeObj.start).isEqual(Point.fromObject(rangeObj.end))) {
      return;
    }
    marker = this.editor.markBufferRange(Range.fromObject(rangeObj), {
      invalidate: 'never'
    });
    this.highlightMap.set(clientId, marker);
    this.editor.decorateMarker(marker, {
      type: 'highlight',
      class: 'selection'
    });
  }

  onSave(relativePath) {
    this.modify( async () => {
      console.log('on save: ');
      console.log(atom.project.getPaths());
      if (typeof this.editor.getPath() !== 'undefined') {
        await this.editor.save();
      }
    });
    // atom.workspace.getActivePane().saveItem(atom.workspace.getActivePaneItem());
    // if (typeof this.editor.getPath() !== 'undefined') {
    //   this.editor.save();
    // } else {
    //   let absolutePath = path.join(atom.project.getPaths()[0], relativePath);
    //   this.editor.saveAs(absolutePath);
    // }
  }

  onInsert(position, text) {
    this.modify( () => {
      let from = this.buffer.positionForCharacterIndex(position);
      this.buffer.insert(from, text);
    })
  }

  onRemove(position, length) {
    this.modify( () => {
      let from = this.buffer.positionForCharacterIndex(position);
      let to = this.buffer.positionForCharacterIndex(position + length);
      this.buffer.delete([from.toArray(), to.toArray()]);
    })
  }

  modify(callback) {
    this.push = false;
    callback();
    this.push = true;
  }

  resetMarker(clientId) {
    this.highlightMap.resetMarker(clientId);
    this.indicatorMap.resetMarker(clientId);
  }

  // TODO more cleanups required
  destroy() {
    this.editor.getDefaultMarkerLayer().clear();
  }
}

Map.prototype.resetMarker = function(clientId) {
  let marker = this.get(clientId);
  if (typeof marker !== 'undefined') {
    marker.destroy();
  }
  return marker;
}
